src/libstd/io/buffered.rs

```diff
@@ -588,7 +588,14 @@ impl<W: Write> BufWriter<W> {
     pub fn into_inner(mut self) -> Result<W, IntoInnerError<BufWriter<W>>> {
         match self.flush_buf() {
             Err(e) => Err(IntoInnerError(self, e)),
-            Ok(()) => Ok(self.inner.take().unwrap())
+            Ok(()) => {
+                use crate::mem::forget;
+                use crate::ptr::{drop_in_place, read};
+                let inner = unsafe { read(&mut self.inner) };
+                unsafe { drop_in_place(&mut self.buf) };
```

Comment:
I notice that if you have a global memory allocator that can panic in dealloc, a panic here would turn into a double-free of `inner` and `self.inner`. The GlobalAlloc trait is currently documented as:

> - It's undefined behavior if global allocators unwind. This restriction may be lifted in the future, but currently a panic from any of these functions may lead to memory unsafety.

so this seems fine for now but needs at least a comment why the code is safe despite appearances.